crates/core/src/kernel/models/actions.rs

```diff
@@ -134,10 +148,72 @@ pub fn contains_timestampntz<'a>(mut fields: impl Iterator<Item = &'a StructFiel
     fields.any(|f| _check_type(f.data_type()))
 }
 
+pub(crate) trait ProtocolExt {
+    fn reader_features_set(&self) -> Option<HashSet<ReaderFeature>>;
+    fn writer_features_set(&self) -> Option<HashSet<WriterFeature>>;
+    fn append_reader_features(self, reader_features: &[ReaderFeature]) -> Protocol;
+    fn append_writer_features(self, writer_features: &[WriterFeature]) -> Protocol;
+    fn move_table_properties_into_features(
+        self,
+        configuration: &HashMap<String, String>,
+    ) -> Protocol;
+    fn apply_column_metadata_to_protocol(self, schema: &StructType) -> DeltaResult<Protocol>;
+    fn apply_properties_to_protocol(
+        self,
+        new_properties: &HashMap<String, String>,
+        raise_if_not_exists: bool,
+    ) -> DeltaResult<Protocol>;
+}
+
+impl ProtocolExt for Protocol {
+    fn reader_features_set(&self) -> Option<HashSet<ReaderFeature>> {
+        self.reader_features()
+            .map(|features| features.iter().cloned().collect())
+    }
+
+    fn writer_features_set(&self) -> Option<HashSet<WriterFeature>> {
+        self.writer_features()
+            .map(|features| features.iter().cloned().collect())
+    }
+
+    fn append_reader_features(self, reader_features: &[ReaderFeature]) -> Protocol {
+        let mut inner = ProtocolInner::from_kernel(&self);
+        inner = inner.append_reader_features(reader_features.iter().cloned());
+        inner.as_kernel()
+    }
+    fn append_writer_features(self, writer_features: &[WriterFeature]) -> Protocol {
+        let mut inner = ProtocolInner::from_kernel(&self);
+        inner = inner.append_writer_features(writer_features.iter().cloned());
+        inner.as_kernel()
+    }
+    fn move_table_properties_into_features(
+        self,
+        configuration: &HashMap<String, String>,
+    ) -> Protocol {
+        let mut inner = ProtocolInner::from_kernel(&self);
+        inner = inner.move_table_properties_into_features(configuration);
+        inner.as_kernel()
+    }
+    fn apply_column_metadata_to_protocol(self, schema: &StructType) -> DeltaResult<Protocol> {
+        let mut inner = ProtocolInner::from_kernel(&self);
+        inner = inner.apply_column_metadata_to_protocol(schema)?;
+        Ok(inner.as_kernel())
+    }
+    fn apply_properties_to_protocol(
+        self,
+        new_properties: &HashMap<String, String>,
+        raise_if_not_exists: bool,
+    ) -> DeltaResult<Protocol> {
+        let mut inner = ProtocolInner::from_kernel(&self);
+        inner = inner.apply_properties_to_protocol(new_properties, raise_if_not_exists)?;
+        Ok(inner.as_kernel())
+    }
+}
+
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
 #[serde(rename_all = "camelCase")]
 /// Defines a protocol action
-pub struct Protocol {
+pub(crate) struct ProtocolInner {
```

Comment:
Will you please add some rustdoc here on the purpose of `ProtocolInner`, I presume it's a temporary shim struct that goes away in the future.

Reply:
added some docs and also directly deprecated the struct (and similar ones). 

Getting rid of all deprecated call sites will be fun after kernel migration :) - there is quite a few by now 😆.